rts/Lua/LuaVAOImpl.cpp

```diff
@@ -185,6 +194,116 @@ int LuaVAOImpl::AddObjectsToSubmissionImpl(const sol::stack_table& ids)
+	std::vector<SDrawElementsIndirectCommand>& submitCmds)
+{
+	const size_t removedObjectCount = (removedCount.has_value()? removedCount.value() : removedObjects.size());
+	const size_t addedObjectCount = (addedCount.has_value()? addedCount.value() : addedObjects.size());
+	if (removedObjectCount == 0 && addedObjectCount == 0) return;
+
+	size_t firstChangedBinIndex = -1;
+
+	for (std::size_t i = 1; i <= removedObjectCount; ++i) {
+		lua_Number objIdLua = removedObjects.raw_get<lua_Number>(i);
+		const int objId = spring::SafeCast<int, lua_Number>(objIdLua);
+
+		const TObj* obj = LuaUtils::SolIdToObject<TObj>(objId, __func__);
+		const int modelId = obj->model->id;
+		const size_t binIndex = modelIdToBinIndex[modelId];
+		firstChangedBinIndex = std::min(firstChangedBinIndex, binIndex);
+		auto& bin = bins[binIndex];
+
+		if (bin.objIds.size() == 1) {
+			modelIdToBinIndex[bins.back().modelId] = binIndex;
+			modelIdToBinIndex.erase(modelId);
+			objIdToLocalInstance.erase(objId);
+			bins[binIndex] = bins.back();
+			bins.pop_back();
+			submitCmds[binIndex] = submitCmds.back();
+			submitCmds.pop_back();
+			continue;
+		}
+
+		size_t localInstance = objIdToLocalInstance[objId];
+		bin.instanceData[localInstance] = bin.instanceData.back();
+		bin.instanceData.pop_back();
+		objIdToLocalInstance[bin.objIds.back()] = localInstance;
+		objIdToLocalInstance.erase(objId);
+		bin.objIds[localInstance] = bin.objIds.back();
+		bin.objIds.pop_back();
+		--submitCmds[binIndex].instanceCount;
+	}
+
+	for (std::size_t i = 1; i <= addedObjectCount; ++i) {
+		lua_Number objIdLua = addedObjects.raw_get<lua_Number>(i);
+		const int objId = spring::SafeCast<int, lua_Number>(objIdLua);
+
+		const TObj* obj = LuaUtils::SolIdToObject<TObj>(objId, __func__);
+		const auto model = obj->model;
+		const int modelId = model->id;
+
+		auto binIndexIt = modelIdToBinIndex.find(modelId);
+		size_t binIndex;
+
+		if (binIndexIt == modelIdToBinIndex.end()) {
+			binIndex = bins.size();
+			modelIdToBinIndex[modelId] = binIndex;
+			bins.emplace_back(modelId);
+			submitCmds.emplace_back(
+				model->indxCount,
+				0u,
+				model->indxStart,
+				0u,
+				0u
+			);
+		} else {
+			binIndex = binIndexIt->second;
+		}
+
+		firstChangedBinIndex = std::min(firstChangedBinIndex, binIndex);
+		auto& bin = bins[binIndex];
+
+		objIdToLocalInstance[objId] = bin.objIds.size();
+		bin.objIds.emplace_back(objId);
+		const auto matIndex = matrixUploader.GetElemOffset(obj);
+		const auto uniIndex = modelsUniformsStorage.GetObjOffset(obj);
+		const size_t bposeIndex = matrixUploader.GetElemOffset(model);
+		bin.instanceData.emplace_back(
+			static_cast<uint32_t>(matIndex),
+			static_cast<uint8_t>(obj->team), // initial team id, will not be updated (unreliable: shader must obtain it somehow else, otherwise it will not reflect transfer to another team)
+			static_cast<uint8_t>(0), // no draw flags, they will not be updated
```

Comment:
Please still submit initial drawflags in order to adhere to the existing conventions.